src/test/csf/collectors.h

```diff
@@ -0,0 +1,740 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012-2017 Ripple Labs Inc
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+#ifndef RIPPLE_TEST_CSF_COLLECTORS_H_INCLUDED
+#define RIPPLE_TEST_CSF_COLLECTORS_H_INCLUDED
+
+#include <ripple/basics/UnorderedContainers.h>
+#include <boost/optional.hpp>
+#include <chrono>
+#include <ostream>
+#include <test/csf/Histogram.h>
+#include <test/csf/SimTime.h>
+#include <test/csf/events.h>
+#include <tuple>
+
+namespace ripple {
+namespace test {
+namespace csf {
+
+//  A collector is any class that implements
+//
+//      on(NodeID, SimTime, Event)
+//
+//  for all events emitted by a Peer.
+//
+//  This file contains helper functions for composing different collectors
+//  and also defines several standard collectors available for simulations.
+
+
+/** Group of collectors.
+
+    Presents a group of collectors as a single collector which process an event
+    by calling each collector sequentially. This is analagous to CollectorRefs
+    in CollectorRef.h, but does *not* erase the type information of the combined
+    collectors.
```

Comment:
Yes, I still have mixed feelings. I don't like non-type-erased everywhere, because it seemed to pervade lots of other types. However, with a type-erased version, there was a point when I profiled a large simulation (some flavor of `ScaleFreeSim`) and this did matter.  Maybe if I was more careful about compiler flags it could've devirtualized things.